buildFlutterApplication: Wrap buildDartApplication

hacker1024 opened the pull request:
###### Description of changes

This PR adds all shared missing features from `buildFlutterApplication` to `buildDartApplication`, and re-implements `buildFlutterApplication` as a wrapper on `buildDartApplication`.

Most notably, it adds the package override mechanism made during the development of `buildFlutterApplication` to `buildDartApplication`. This feature is very useful for FFI dependencies on pure Dart applications.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Misterio77 commented:
Hey!

May I ask why the deps list is required? My understanding was that fetching them being a FOD means that `depsHash` is everything that's needed, and that flutter apps need them because the sdk has additional stuff.

hacker1024 commented:
> Hey!
> 
> May I ask why the deps list is required? My understanding was that fetching them being a FOD means that `depsHash` is everything that's needed, but that flutter apps need them because the sdk has additional stuff.

The deps list is used by `buildDartApplication` to work out which dependencies are in use by the application, and apply their corresponding package overrides.

For example, applications that use [`package:olm`](https://pub.dev/packages/olm) need [the olm library](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/development/libraries/olm/default.nix), so there's a [package override to add it to `runtimeDependencies`](https://github.com/NixOS/nixpkgs/blob/16bac62854a283e61f091af1712e169a3ae7e42f/pkgs/development/compilers/dart/package-overrides/olm/default.nix).

This allows Dart and Flutter apps with common dependencies to share derivation properties without repetition.

Misterio77 commented:
That makes sense. Thanks!

hacker1024 commented:
Rebased on master due to Dart and Flutter upgrades.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2230

lelgenio commented:
Could we maybe derive the contents of deps.json from pubspec.lock if it's present? Not that it would be too much of an inconvenience to update both files.

hacker1024 commented:
> Could we maybe derive the contents of deps.json from pubspec.lock if it's present? Not that it would be too much of an inconvenience to update both files.

The JSON dependency list is used purely during evaluation, so this is not possible without import-from-derivation.

SuperSandro2000 commented on pkgs/build-support/dart/build-dart-application/default.nix:
```diff
@@ -32,35 +37,80 @@
+      jq
+    ];
+
+    preUnpack = ''
+      ${lib.optionalString (!autoDepsList) ''
+        if ! { [ '${lib.boolToString (depsListFile != null)}' = 'true' ] ${lib.optionalString (depsListFile != null) "&& cmp -s <(jq -Sc . '${depsListFile}') <(jq -Sc . '${finalAttrs.passthru.dartDeps.depsListFile}')"}; }; then
+          echo 1>&2 -e '\nThe dependency list file was either not given or differs from the expected result.' \
+                      '\nPlease choose one of the following solutions:' \
+                      '\n - Duplicate the following file and pass it to the depsListFile argument.' \
+                      '\n   ${finalAttrs.passthru.dartDeps.depsListFile}' \
+                      '\n - Set autoDepsList to true (not supported by Hydra or permitted in Nixpkgs)'.
+          exit 1
+        fi
+      ''}
```
```suggestion
    preUnpack = lib.optionalString (!autoDepsList) ''
      if ! { [ '${lib.boolToString (depsListFile != null)}' = 'true' ] ${lib.optionalString (depsListFile != null) "&& cmp -s <(jq -Sc . '${depsListFile}') <(jq -Sc . '${finalAttrs.passthru.dartDeps.depsListFile}')"}; }; then
        echo 1>&2 -e '\nThe dependency list file was either not given or differs from the expected result.' \
                    '\nPlease choose one of the following solutions:' \
                    '\n - Duplicate the following file and pass it to the depsListFile argument.' \
                    '\n   ${finalAttrs.passthru.dartDeps.depsListFile}' \
                    '\n - Set autoDepsList to true (not supported by Hydra or permitted in Nixpkgs)'.
        exit 1
      fi
    '' +
```

SuperSandro2000 commented on pkgs/build-support/dart/build-dart-application/default.nix:
```diff
@@ -32,35 +37,80 @@
-  # When stripping, it seems some ELF information is lost and the dart VM cli
-  # runs instead of the expected program. Don't strip if it's an exe output.
-  dontStrip = args.dontStrip or (dartOutputType == "exe");
+      ${args.preUnpack or ""}
+    '';
```
```suggestion
    '' + args.preUnpack or "";
```

hacker1024 replied:
Is there a performance reason for this change, or an official style specification? Or is it just personal preference?

SuperSandro2000 replied:
The code is simpler this way and if we refactor things we might as well simplify them.

SuperSandro2000 commented on pkgs/build-support/dart/build-dart-application/hooks/dart-fixup-hook.sh:
```diff
@@ -0,0 +1,21 @@
+        # Add runtime library dependencies to the LD_LIBRARY_PATH.
+        # For some reason, the RUNPATH of the executable is not used to load dynamic libraries in dart:ffi with DynamicLibrary.open().
+        #
+        # This could alternatively be fixed with patchelf --add-needed, but this would cause all the libraries to be opened immediately,
+        # which is not what application authors expect.
```
but what we should probably do because LD_LIBRARY_PATH taints all programs start from the wrapped program and potentially causing them to load the wrong libraries

hacker1024 replied:
This can be done in a later PR - this one is just a refactor.

hacker1024 commented:
Rebased on master.

hacker1024 commented on pkgs/build-support/flutter/default.nix:
```diff
@@ -29,131 +17,54 @@ let
 
-      runHook postInstall
-    '';
+    for f in $(find $out/app -iname "*.desktop" -type f); do
```
I think the desktop file search is alright, though I'm not sure why it's there to begin with.

As for the main binary search, determining the name of the executable in advance is non-trivial. We could use the `depsListFile` to find the `root` package, but this would involve a linear search and be quite inefficient. We could parse the `pubspec.yaml`, but this would require a YAML parser to be added to the build closure. I don't think it's very likely for this `find` invocation to go wrong, as I've never seen a case where other files are added to the `app` directory.

The third `find` cannot practically be done in another way, as the names of library files added by plugins are quite arbitrary.

@mkg20001 I have some questions:
- Why search for desktop files here? Is there a tool in the Flutter ecosystem that puts them in this location? If not, we can probably remove this in favor of `copyDesktopItems`.
- Why do the library files need to be executable?

Anyway, none of this is new in this PR, which is mainly a refactor. I think we can merge this PR first, and change the behavior in question later.

mkg20001 replied:
ah totally missed this. most of this is hackery from my first packaging attempts. try to remove some and see if things break.

library files need to be executable, but some weren't. maybe that is fixed now.

as for the ln that's mostly duct tape.

desktop files: i think was firmware-updater that puts them there

hacker1024 commented on pkgs/build-support/flutter/default.nix:
```diff
@@ -29,131 +17,54 @@ let
-  });
+    # remove stuff like /build/source/packages/ubuntu_desktop_installer/linux/flutter/ephemeral
+    for f in $(find $out/app -executable -type f); do
+      if patchelf --print-rpath "$f" | grep /build; then # this ignores static libs (e,g. libapp.so) also
```
What is brittle? The hardcoded `/build` or the method in general? If the latter, how else can we work around this issue? Ideally it should be fixed upstream.

I will make the suggested change in another PR, as I think it is beyond the scope of this one, which is mainly a refactor.

hacker1024 commented:
@SuperSandro2000 Thanks for the detailed review. While you raise many good points, I think some are beyond the scope of this PR, as the code in question is not being changed here.

More PRs should certainly be made to fix the existing issues, but as this PR is mainly a refactor, I think they can be left alone for now.

hacker1024 commented:
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>dart-sass</li>
    <li>expidus.file-manager</li>
    <li>firmware-updater</li>
    <li>fluffychat</li>
    <li>protoc-gen-dart</li>
    <li>shopware-cli</li>
    <li>yubioath-flutter</li>
  </ul>
</details>

zimbatm commented:
@hacker1024 sorry, it looks like pkgs/development/tools/misc/dart-sass/default.nix was modified in the meantime

hacker1024 commented:
Oof! Rebased again.

hacker1024 commented:
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>dart-sass</li>
    <li>expidus.file-manager</li>
    <li>firmware-updater</li>
    <li>fluffychat</li>
    <li>protoc-gen-dart</li>
    <li>shopware-cli</li>
    <li>yubioath-flutter</li>
  </ul>
</details>

hacker1024 commented:
Google has made some server-side changes that cause SSL issues. This has been fixed in the upcoming Flutter 3.13.0 PR, but that is a WIP and I'd prefer that this was merged first (it's easier to rebase on this than the other way round).

To fix the issue for review, the fixes for [Flutter](https://github.com/NixOS/nixpkgs/pull/251409/commits/a65834c1e6aee803d731b2e84bfab5d7e65fd9ea) and [Dart](https://github.com/hacker1024/nixpkgs/commit/b3594ce1ecb46f32027278f2e90d56f2ed359c57) can be cherry-picked.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2599

mkg20001 commented:
Needs rebase.

hacker1024 committed:
buildDartApplication: Format with nixpkgs-fmt

hacker1024 committed:
buildDartApplication: Allow supplying runtime dependencies

hacker1024 committed:
fetchDartDeps: Disable install phase in depsListDrv

hacker1024 committed:
buildDartApplication: Use package override mechanism from buildFlutterApplication

This allows Dart applications to benefit from the package override system, which is useful for things like FFI dependencies.

hacker1024 committed:
buildDartApplication: Allow passing a `sdkSetupScript`

hacker1024 committed:
buildDartApplication: Recognise extraWrapProgramArgs

hacker1024 committed:
buildFlutterApplication: Wrap buildDartApplication

hacker1024 commented:
Rebased. I needed to update the Dart fixup hook to work better with the recently added use of wrapGAppsHook in Flutter, but everything else is pretty much the same.

hacker1024 committed:
dartHooks.dartFixupHook: Add trailing newline

hacker1024 commented:
Result of `nixpkgs-review pr 231483` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>20 packages built:</summary>
  <ul>
    <li>dart-sass</li>
    <li>discourse</li>
    <li>discourseAllPlugins</li>
    <li>domine</li>
    <li>expidus.file-manager</li>
    <li>firmware-updater</li>
    <li>fluffychat</li>
    <li>protoc-gen-dart</li>
    <li>python310Packages.nbdev</li>
    <li>python310Packages.nbdev.dist</li>
    <li>python311Packages.nbdev</li>
    <li>python311Packages.nbdev.dist</li>
    <li>quarto</li>
    <li>quartoMinimal</li>
    <li>rstudio</li>
    <li>rstudio-server</li>
    <li>rstudioServerWrapper</li>
    <li>rstudioWrapper</li>
    <li>shopware-cli</li>
    <li>yubioath-flutter</li>
  </ul>
</details>
